File: src/minigames/guess_it/src/main/java/work/lclpnet/ap2/game/guess_it/data/InputValue.java
package work.lclpnet.ap2.game.guess_it.data;

import it.unimi.dsi.fastutil.Pair;
import lombok.Getter;
import net.minecraft.server.level.ServerPlayer;
import org.jetbrains.annotations.Nullable;
import work.lclpnet.kibu.translate.Translations;
import work.lclpnet.kibu.translate.text.TranslatedText;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.function.Function;

import static java.lang.Math.clamp;
import static net.minecraft.ChatFormatting.RED;
import static net.minecraft.ChatFormatting.YELLOW;
import static work.lclpnet.kibu.translate.text.FormatWrapper.styled;

@SuppressWarnings("UnusedReturnValue")
public class InputValue {

    private final List<InputRule> rules = new ArrayList<>();
    @Getter
    private boolean once = false;

    public InputValue validate(InputParser validator, Function<String, TranslatedText> errorMessage) {
        rules.add(new InputRule(validator, errorMessage));
        return this;
    }

    public InputValue validateInt(Translations translations) {
        return validate(InputValue::intValue, input ->
                translations.translateText("game.ap2.guess_it.input.int", styled(input, YELLOW)).formatted(RED));
    }

    public InputValue validateFloat(Translations translations, int precision) {
        return validate((input, player) -> floatValue(input, player, translations, precision), input ->
                translations.translateText("game.ap2.guess_it.input.float", styled(input, YELLOW)).formatted(RED));
    }

    public InputValue onlyOnce() {
        once = true;
        return this;
    }

    public Pair<String, @Nullable TranslatedText> validate(String input, ServerPlayer player) {
        for (InputRule rule : rules) {
            var res = rule.parser().parse(input, player);

            if (res.isEmpty()) {
                return Pair.of(input, rule.errorMessage.apply(input));
            }

            input = res.get();
        }

        return Pair.of(input, null);
    }

    private static Optional<String> floatValue(String s, ServerPlayer player, Translations translations, int precision) {
        s = s.replace(',', '.');

        float f;

        try {
            f = Float.parseFloat(s);
        } catch (NumberFormatException _) {
            return Optional.empty();
        }

        String fmt = "%." + clamp(precision, 0, 7) + "f";
        Locale locale = translations.getLocale(player);
        String str = String.format(locale, fmt, f);

        return Optional.of(str);
    }

    public static Optional<String> intValue(String s, ServerPlayer player) {
        try {
            int i = Integer.parseInt(s, 10);
            return Optional.of(String.valueOf(i));
        } catch (NumberFormatException _) {
            return Optional.empty();
        }
    }

    public interface InputParser {
        Optional<String> parse(String input, ServerPlayer player);
    }

    private record InputRule(InputParser parser, Function<String, TranslatedText> errorMessage) {}
}

File: src/minigames/guess_it/src/main/kotlin/work/lclpnet/ap2/game/guess_it/data/InputValue.kt
package work.lclpnet.ap2.game.guess_it.data

import net.minecraft.ChatFormatting
import net.minecraft.server.level.ServerPlayer
import work.lclpnet.kibu.translate.Translations
import work.lclpnet.kibu.translate.text.FormatWrapper
import work.lclpnet.kibu.translate.text.TranslatedText
import java.util.function.Function

class InputValue {

    private val rules = ArrayList<InputRule>()

    var once = false
        private set

    fun validate(validator: InputParser, errorMessage: Function<String, TranslatedText>): InputValue {
        rules.add(InputRule(validator, errorMessage))
        return this
    }

    fun validateInt(translations: Translations): InputValue = validate(
        { s, _ -> intValue(s) },
        { input ->
            translations.translateText(
                "input.int",
                FormatWrapper.styled(input, ChatFormatting.YELLOW)
            ).withStyle(ChatFormatting.RED)
        }
    )

    fun validateFloat(translations: Translations, precision: Int): InputValue = validate(
        { input, player ->
            floatValue(
                input,
                player,
                translations,
                precision
            )
        },
        { input ->
            translations.translateText(
                "input.float",
                FormatWrapper.styled(input, ChatFormatting.YELLOW)
            ).withStyle(ChatFormatting.RED)
        }
    )

    fun onlyOnce(): InputValue {
        once = true
        return this
    }

    fun validate(input: String, player: ServerPlayer): Pair<String?, TranslatedText?> {
        var input = input

        for (rule in rules) {
            val res = rule.parser.parse(input, player) ?: return input to rule.errorMessage.apply(input)

            input = res
        }

        return input to null
    }

    fun interface InputParser {
        fun parse(input: String, player: ServerPlayer): String?
    }

    private data class InputRule(
        val parser: InputParser,
        val errorMessage: Function<String, TranslatedText>
    )

    companion object {
        private fun floatValue(s: String, player: ServerPlayer, translations: Translations, precision: Int): String? {
            val s = s.replace(',', '.')
            val f = s.toFloatOrNull() ?: return null

            val fmt = "%." + Math.clamp(precision.toLong(), 0, 7) + "f"
            val locale = translations.getLocale(player)

            return String.format(locale, fmt, f)
        }

        fun intValue(s: String): String? =
            s.toIntOrNull()?.toString()
    }
}
